From e165cbc76a81e6235de9ce8d2e12c35a260f24d0 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Tue, 09 Mar 2021 16:16:10 +0200
Subject: [PATCH] [SW-34382] boot info: preserve pre-reset data

Linux is storing the pre-reset data in scratchpad sram when there is a
fatal error or reboot request. this data should remain in the sp
sram and used by either preboot or linux for error reporting to
the bmc.

Change-Id: Ide809df1ef0df6bc51aaf30c290c6810a6d3617d
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/include/boot_info.h b/include/boot_info.h
index abf60cf..0779b03 100644
--- a/include/boot_info.h
+++ b/include/boot_info.h
@@ -11,6 +11,7 @@
 
 void boot_info_finalize(void);
 int boot_info_set(u32 type, const void *data, u32 size);
-int boot_info_get(u32 type, void *data, u32 size);
+int boot_info_get(u32 type, void *data, u32 max_data_size,
+		  u32 *actual_data_size);
 
 #endif /* _BOOT_INFO_H_ */
diff --git a/include/preboot.h b/include/preboot.h
index 230ca42..952a92a 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -10,7 +10,8 @@
 
 #include <common.h>
 
-#define ARRAY_SIZE(x)				(sizeof(x) / sizeof((x)[0]))
+#define ARRAY_SIZE(x)		(sizeof(x) / sizeof((x)[0]))
+#define ROUND_UP4(n)		(((n) + sizeof(u32) - 1) & ~(sizeof(u32) - 1))
 
 #ifdef __arc__
 #define wfe()	__asm__ __volatile__("sleep 0x11\n\t" : : )
diff --git a/preboot/boot_info.c b/preboot/boot_info.c
index a5421ec..0bfcd8b 100644
--- a/preboot/boot_info.c
+++ b/preboot/boot_info.c
@@ -16,7 +16,6 @@
 
 #define MAX_INFO_DESCS	5
 
-#define ROUND_UP4(n)	(((n) + sizeof(u32) - 1) & ~(sizeof(u32) - 1))
 #define BOOT_INFO_SZ	(sizeof(struct hl_boot_info_ptrs) + \
 			sizeof(struct hl_boot_info_desc) * MAX_INFO_DESCS)
 
@@ -140,7 +139,8 @@
 	return 0;
 }
 
-int boot_info_get(u32 type, void *data, u32 size)
+int boot_info_get(u32 type, void *data, u32 max_data_size,
+		  u32 *actual_data_size)
 {
 	struct hl_boot_info_ptrs ptrs;
 	struct hl_boot_info_desc desc;
@@ -148,7 +148,7 @@
 	u8 chksum[HL_CHKSUM_SIZE];
 
 	/* Size must be 32bit aligned */
-	if (size & (sizeof(u32) - 1))
+	if (max_data_size & (sizeof(u32) - 1))
 		return 1;
 
 	dst = data;
@@ -169,7 +169,7 @@
 		if (desc.magic != (HL_BOOT_INFO_DESC_MAGIC | type))
 			return 1;
 
-		if (size < desc.size)
+		if (max_data_size < desc.size)
 			return 1;
 
 		src = offset2ptr(desc.offset, BOOT_INFO_BASE);
@@ -180,6 +180,8 @@
 		if (memcmp(chksum, desc.chksum, HL_CHKSUM_SIZE))
 			return 1;
 
+		*actual_data_size = desc.size;
+
 		return 0;
 	}
 
diff --git a/preboot/preboot.c b/preboot/preboot.c
index e299740..134def9 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -7,6 +7,7 @@
 
 #include <stddef.h>
 #include <stdint.h>
+#include <time.h>	/* required for nvme data */
 
 #include <psoc_global_conf_regs.h>
 #include <pcie_dbi_regs.h>
@@ -20,6 +21,8 @@
 #include <boot_info.h>
 #include <hl_boot_desc.h>
 #include <crc32.h>
+#include <rma_if.h>	/* required for nvme data */
+#include <hl_nvme_mi_data.h>
 
 #include "common.h"
 #include "uart.h"
@@ -257,12 +260,19 @@
 	device_id_vendor_id = regmap_readl(mmPCIE_DBI_DEVICE_ID_VENDOR_ID_REG);
 }
 
+/* leaving extra room for future data growth */
+#define DESC_BUFF_SIZE	(2 * sizeof(struct shm_nvme_mi_data))
+
+/* buffer size must be 32bit aligned */
+u8 boot_info_buff[ROUND_UP4(DESC_BUFF_SIZE)];
+
 static void preboot_process_last_boot_info(void)
 {
 	u32 boot_err = HL_FW_FIT_BOOT_ERR_NONE;
+	u32 data_size;
 
 	if (!boot_info_get(HL_BOOT_INFO_FIT_BOOT_ERROR, &boot_err,
-			   sizeof(boot_err))) {
+			   sizeof(boot_err), &data_size)) {
 		switch (boot_err) {
 		case HL_FW_FIT_BOOT_ERR_NONE:
 			break;
@@ -284,8 +294,13 @@
 		}
 
 		i2csi_set_boot_err(boot_err, true);
-		return;
 	}
+
+	/* preserve nvme data descriptor if it's present in the sp sram */
+	if (!boot_info_get(HL_BOOT_INFO_NVME_STATUS, (void *)boot_info_buff,
+			   sizeof(boot_info_buff), &data_size))
+		boot_info_set(HL_BOOT_INFO_NVME_STATUS, (void *)boot_info_buff,
+			      data_size);
 }
 
 __weak void hl_dev_pll_init(void) {}
